### PR TITLE
Add support for (Un)MarshalAminoJSON override

### PR DESCRIPTION
**Pull request by amaury1093**
ref: https://github.com/cosmos/cosmos-sdk/issues/7109#issuecomment-689709943

If a type implements `MarshalAminoJSON` and `UnmarshalAminoJSON`, then amino will use these methods for JSON marshaling.

This is the json-equivalent for methods `MarshalAmino` and `UnmarshalAmino`

also cc @alexanderbez, since I can't add you as a reviewer

### CODE REVIEW

**Comment by alessio** [comment]
> also cc @alexanderbez, since I can't add you as a reviewer

Added. And you should be in the position to add him in future PRs as well :+1:

File: repr_test.go [hunk 1]
```diff
@@ -97,3 +97,54 @@ func TestMarshalAminoJSON(t *testing.T) {
...
+
+type Bar struct {
+	a string
+	b int
+	c []*Bar
+	D string // exposed
+}
+
+func (b Bar) MarshalAminoJSON() ([]pair, error) { // nolint: golint
+	return []pair{
```
Comment (alexanderbez) [inline repr_test.go:108]: Curious, why does this return a list of pairs and not bytes? What is the typical implementation of `MarshalAminoJSON` look like?

Comment (amaury1093) [reply repr_test.go:108]: Yeah, it can return whatever it wants. I just copied the `MarshalAmino` tests, which returned pairs. Can change to []byte, if you prefer that.

In the SDK we'll want to return []byte, like here: https://github.com/cosmos/cosmos-sdk/pull/7276/files#diff-1f6fdece2a5550f27ed0b77247d58f8bR60-R65

**Review by alexanderbez — APPROVED** [review]

File: json-encode.go [hunk 2]
```diff
@@ -61,6 +82,17 @@ func (cdc *Codec) encodeReflectJSON(w io.Writer, info *TypeInfo, rv reflect.Valu
...
+		if rv.Addr().Type().Implements(jsonMarshalerType) {
+			err = invokeMarshalJSON(w, rv.Addr())
+			return
+		}
+	} else if rv.Type().Implements(jsonMarshalerType) {
+		err = invokeMarshalJSON(w, rv)
+		return
+	}
+
+	// Handle override if rv implements MarshalAmino.
```
Comment (aaronc) [inline json-encode.go]: Why is this needed?

File: codec.go [hunk 3]
```diff
@@ -86,10 +86,14 @@ type ConcreteInfo struct {
...
-	IsAminoUnmarshaler     bool         // Implements UnmarshalAmino(<ReprObject>) (error).
-	AminoUnmarshalReprType reflect.Type // <ReprType>
+	IsAminoMarshaler           bool         // Implements MarshalAmino() (<ReprObject>, error).
+	AminoMarshalReprType       reflect.Type // <ReprType>
+	IsAminoUnmarshaler         bool         // Implements UnmarshalAmino(<ReprObject>) (error).
+	AminoUnmarshalReprType     reflect.Type // <ReprType>
+	IsAminoMarshalerJSON       bool         // Implements MarshalAminoJSON() (<ReprObject>, error).
+	AminoMarshalJSONReprType   reflect.Type // <ReprType>
+	IsAminoUnmarshalerJSON     bool         // Implements UnmarshalAminoJSON(<ReprObject>) (error).
+	AminoUnmarshalJSONReprType reflect.Type // <ReprType>
```
Comment (aaronc) [inline codec.go]: nit
```suggestion
	IsAminoJSONMarshaler       bool         // Implements MarshalAminoJSON() (<ReprObject>, error).
	AminoJSONMarshalReprType   reflect.Type // <ReprType>
	IsAminoJSONUnmarshaler     bool         // Implements UnmarshalAminoJSON(<ReprObject>) (error).
	AminoJSONUnmarshalReprType reflect.Type // <ReprType>
```

File: json-decode.go [hunk 4]
```diff
@@ -61,6 +61,29 @@ func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value,
...
+		}
+		// Then, decode from repr instance.
+		uwrm := rv.Addr().MethodByName("UnmarshalAminoJSON")
+		uwouts := uwrm.Call([]reflect.Value{rrv})
+		erri := uwouts[0].Interface()
+		if erri != nil {
+			err = erri.(error)
+		}
+		return
+	}
```
Comment (aaronc) [inline json-decode.go]: Why can't we just do:
```go
	// Handle override if a pointer to rv implements json.Unmarshaler.
	if rv.Addr().Type().Implements(jsonUnmarshalerType) {
		err = rv.Addr().Interface().(AminoJSONUnmarshaler).UnmarshalAminoJSON(bz)
		return
	}
```

Is this because we want to allow any type to be used instead of `[]byte`? Is it better to be consistent with `Un/MarshalAmino` or should we just do the simpler thing and use the same signatures as `Un/MarshalJSON`?

File: json-encode.go [hunk 5]
```diff
@@ -49,6 +49,27 @@ func (cdc *Codec) encodeReflectJSON(w io.Writer, info *TypeInfo, rv reflect.Valu
...
+			return
+		}
+		rinfo, err = cdc.getTypeInfoWlock(info.AminoMarshalJSONReprType)
+		if err != nil {
+			return
+		}
+		// Then, encode the repr instance.
+		err = cdc.encodeReflectJSON(w, rinfo, rrv, fopts)
+		return
+	}
```
Comment (aaronc) [inline json-encode.go]: ditto

Comment (aaronc) [inline json-encode.go]: Ditto, can't we just cast the type and call `MarshalAminoJSON`?

File: json-encode.go [hunk 6]
```diff
@@ -61,6 +82,17 @@ func (cdc *Codec) encodeReflectJSON(w io.Writer, info *TypeInfo, rv reflect.Valu
...
 	// Handle override if rv implements json.Marshaler.
+	if rv.CanAddr() { // Try pointer first.
+		if rv.Addr().Type().Implements(jsonMarshalerType) {
+			err = invokeMarshalJSON(w, rv.Addr())
+			return
+		}
+	} else if rv.Type().Implements(jsonMarshalerType) {
+		err = invokeMarshalJSON(w, rv)
+		return
+	}
```
Comment (amaury1093) [inline json-encode.go]: note to self: to be removed

File: json-decode.go (hunk 4 above)
Comment (amaury1093) [reply json-decode.go]: > Is it better to be consistent with Un/MarshalAmino or should we just do the simpler thing and use the same signatures as Un/MarshalJSON

Yes, that's the whole debate. The way I went for is to keep the symmetry in function signature between:
- Un/MarshalJSON and Un/MarshalBinary
- Un/MarshalAminoJSON and Un/MarshalAmino

File: json-encode.go (hunk 2 above)
Comment (amaury1093) [reply json-encode.go]: removed, a stray copy-paste.

**Review by aaronc — APPROVED** [review]
utACK

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,14 @@
 # Changelog
 
+## 0.15.1 (September 10, 2020)
```
Comment (aaronc) [inline CHANGELOG.md]: `0.15.1` is already tagged on a different branch: https://github.com/tendermint/go-amino/blob/release/v0.15.1/CHANGELOG.md. I think we should put these changes in `0.16.0`. It appears that there are also some changes in `0.15.1` that haven't been merged back into master. @marbar3778 you did the last point release. Are you able to give us some guidance on doing a new release here?

**Review by aaronc — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 7 above)
Comment (tac0turtle) [reply CHANGELOG.md]: I believe the best approach would be to branch off of the most recent release branch or tag. It is unclear if master is working correctly, I vaguely remember there being a bug or two on master that were never fixed.

Comment (amaury1093) [reply CHANGELOG.md]: Done. There's a changelog entry indeed for 0.15.1, but only in this release branch. I added a 0.16.0 one.

Comment (tac0turtle) [reply CHANGELOG.md]: Sorry for not being descriptive. Could you create a branch from release/0.15.1 like rc0/0.16 then put this PR against that?

Comment (amaury1093) [reply CHANGELOG.md]: Ah ok, understood.

**Review by aaronc — APPROVED** [review]
